### ISSUE TITLE
Custom Loss Functions

### ISSUE THREAD

**Issue by raddy** [dmlc/xgboost#353]
So I'm trying to generate a cost function such that the the total sum of scores for a 'go/no-go' set of predictions is maximized. So, zero is no pull, 1 is pull, and the goal is to maximize the sum reward across all pulls. 

So the scoring function is, modeled after the example:
    
    def eval(preds,dtrain):
        truth = dtrain.get_label()
        return 'error', np.sum((preds>0.5).astype(int)*truth))*-1

The objective function, I thought, would look like the following, but perhaps I'm misunderstanding.

    def obj(preds,dtrain):
        labels = dtrain.get_label()
        preds = np.round(np.clip(0,1))
        grad = (preds>.5) * labels * -1
        hess = (preds>.5) * -1
        return grad,hess

So in this example, the labels are a set of floats (scores) with the thought being to directly optimize for this score. However, there's something clearly wrong in this setup, as the results are..well, not exactly right.

**Comment by tqchen** [comment]
This was because of the function you are trying to optimize was not very typical, and second order gradient contain negative values(maybe you want to check if that is correct). Normally hess will have to be positive value(for most convex functions). 

What you are trying to implement was a zero-one loss, where you can not simply take gradient to the input (in most places all of them are zero), This loss function is not smooth, continuous and usually cannot be optimized. I would suggest you try a loss function with better property. For example, logistic loss

**Comment by raddy** [comment]
Well, it's similar to weighted zero-one loss, no? (Where the weight was the original numerical score, but it's only applied if the prediction was a one)

I could rescale the numerical scores to being all positive, and perhaps inverting my prediction, but the hessian still wouldn't be positive. 

Optimizing without using the numerical info just fitting a logistic isn't viable. (This actually does worse than regressing with MLE/MAE directly to the numerical cost, which in turn, does worse than a few hand chosen rules)

**Comment by tqchen** [comment]
You can directly add weight to logistic loss. 
In your zero one loss, all the gradient and second order gradient are zero instead of what you calculated. Such loss function was in nature hard to optimize and usually what people do is optimize a surrogate loss instead. I cannot comment on the performance of the model though, this will depend on how many data and feature you have in your setting
